nibabel/cifti2/cifti2_axes.py

```diff
@@ -0,0 +1,1224 @@
+            matrix.append(mim)
+    return cifti2.Cifti2Header(matrix)
+
+
+class Axis(object):
+    """
+    Generic object describing the rows or columns of a CIFTI vector/matrix
+
+    Attributes
+    ----------
+    arr : np.ndarray
+        (N, ) typed array with the actual information on each row/column
+    """
+    _use_dtype = None
+    arr = None
+
+    def __init__(self, arr):
+        self.arr = np.asarray(arr, dtype=self._use_dtype)
+
+    def get_element(self, index):
+        """
+        Extracts a single element from the axis
+
+        Parameters
+        ----------
+        index : int
+            Indexes the row/column of interest
+
+        Returns
+        -------
+        Description of the row/column
+        """
+        return self.arr[index]
+
+    def __getitem__(self, item):
+        if isinstance(item, int):
+            return self.get_element(item)
+        if isinstance(item, string_types):
+            raise IndexError("Can not index an Axis with a string (except for Parcels)")
+        return type(self)(self.arr[item])
+
+    @property
+    def size(self, ):
+        return self.arr.size
+
+    def __len__(self):
+        return self.size
+
+    def __eq__(self, other):
+        return (type(self) == type(other) and
+                len(self) == len(other) and
+                (self.arr == other.arr).all())
+
+    def __add__(self, other):
+        """
+        Concatenates two Axes of the same type
+
+        Parameters
+        ----------
+        other : Axis
+            axis to be appended to the current one
+
+        Returns
+        -------
+        Axis of the same subtype as self and other
+        """
+        if type(self) == type(other):
+            return type(self)(np.append(self.arr, other.arr))
+        return NotImplemented
+
+
+class BrainModel(Axis):
+    """
+    Each row/column in the CIFTI vector/matrix represents a single vertex or voxel
+
+    This Axis describes which vertex/voxel is represented by each row/column.
+
+    Attributes
+    ----------
+    voxel : np.ndarray
+        (N, 3) array with the voxel indices
+    vertex :  np.ndarray
+        (N, ) array with the vertex indices
+    name : np.ndarray
+        (N, ) array with the brain structure objects
+    """
+    _use_dtype = np.dtype([('vertex', 'i4'), ('voxel', ('i4', 3)),
+                           ('name', 'U%i' % max(len(name) for name in cifti2.CIFTI_BRAIN_STRUCTURES))])
+    _affine = None
+    _volume_shape = None
+
+    def __init__(self, arr, affine=None, volume_shape=None, nvertices=None):
+        """
+        Creates a new BrainModel axis defining the vertices and voxels represented by each row/column
+
+        Parameters
+        ----------
+        arr : np.ndarray
+            (N, ) structured array with for every element a tuple with 3 elements:
+            - vertex index (-1 for voxels)
+            - 3 voxel indices (-1 for vertices)
+            - string (name of brain structure)
```

Comment:
@MichielCottaar I think that passing the voxel/vertex/structure elements separated as in the attributes presented on line 130 might be an easier to understand interface. The typed arrays with `-1` harcoded for an ignored value seems suboptimal as interface design. @effigies , you have more the hand on what's the interface style in nibabel, what do you think?

Alternatively, this constructor could be hidden and just creating the objects from the factory methods, such as `from_mask`, which have very comfy and clear interfaces, might be a good option.